Defer closing of the database when an error occurred, results in a panic.

**Version being used**

```
ajfs: 0.0.1-beta5 cee39fc9d38b47cf9ae1a43f61e6a194042bc6c6
```

**Describe the bug**

I was trying to scan an attached SSD on my Mac and it resulted in the following panic.

```
panic: FinishEntries was never called

goroutine 1 [running]:
github.com/andrejacobs/ajfs/internal/db.(*DatabaseFile).finishCreation(0x1400008e780?)
	/home/runner/work/ajfs/ajfs/internal/db/db.go:545 +0x2c0
github.com/andrejacobs/ajfs/internal/db.(*DatabaseFile).Close(0x140000ec3c0)
	/home/runner/work/ajfs/ajfs/internal/db/db.go:252 +0x54
github.com/andrejacobs/ajfs/internal/app/scan.Run.func1()
	/home/runner/work/ajfs/ajfs/internal/app/scan/scan.go:108 +0x24
github.com/andrejacobs/ajfs/internal/app/scan.Run({{{0x16daa3940, 0x30}, 0x0, 0x1, {0x10286df08, 0x14000094018}, {0x10286df08, 0x14000094020}}, {0x102869858, 0x102869858, ...}, ...})
	/home/runner/work/ajfs/ajfs/internal/app/scan/scan.go:172 +0x748
github.com/andrejacobs/ajfs/cmd/ajfs/commands.init.func9(0x140000f6700?, {0x140000ea2d0, 0x2, 0x1026da4d9?})
	/home/runner/work/ajfs/ajfs/cmd/ajfs/commands/scan.go:138 +0x180
github.com/spf13/cobra.(*Command).execute(0x102b81260, {0x140000ea280, 0x5, 0x5})
	/home/runner/work/ajfs/ajfs/vendor/github.com/spf13/cobra/command.go:1019 +0x7bc
github.com/spf13/cobra.(*Command).ExecuteC(0x102b7f880)
	/home/runner/work/ajfs/ajfs/vendor/github.com/spf13/cobra/command.go:1148 +0x350
github.com/spf13/cobra.(*Command).Execute(...)
	/home/runner/work/ajfs/ajfs/vendor/github.com/spf13/cobra/command.go:1071
github.com/andrejacobs/ajfs/cmd/ajfs/commands.Execute()
	/home/runner/work/ajfs/ajfs/cmd/ajfs/commands/root.go:55 +0x24
main.main()
	/home/runner/work/ajfs/ajfs/cmd/ajfs/main.go:30 +0x1c
```

This appears to happen when an error was encountered (maybe you don't have read access etc.), the function calls return err and then the deferred close gets called with the panic that the file wasn't written correctly.

**To Reproduce**
See above.

**Expected behavior**
The original error should be reported back to the user and the file should still be closed but without a panic.

**Desktop (please complete the following information):**

- OS: macOS

--- a/.github/workflows/makefile.yml
+++ b/.github/workflows/makefile.yml
@@ -19,7 +19,7 @@ jobs:
             - name: Set up Go
               uses: actions/setup-go@v5
               with:
-                  go-version: stable
+                  go-version: 1.25
 
             - name: Install dependencies
               run: make go-deps
@@ -30,7 +30,7 @@ jobs:
             - name: Checking code quality with golangci-lint
               uses: golangci/golangci-lint-action@v8
               with:
-                  version: v2.4.0
+                  version: latest
 
             - name: Checking copyright and license is applied
               run: >-

--- a/cmd/ajfs/commands/filter.go
+++ b/cmd/ajfs/commands/filter.go
@@ -22,6 +22,7 @@ package commands
 
 import (
 	"fmt"
+	"runtime"
 
 	"github.com/andrejacobs/ajfs/internal/app/config"
 	"github.com/andrejacobs/ajfs/internal/filter"
@@ -69,5 +70,11 @@ func parseFilterConfig() (*config.FilterConfig, error) {
 
 	result.FileExcluder = file.MatchAppleDSStore(exclF)
 	result.DirExcluder = exclD
+
+	if runtime.GOOS == "darwin" {
+		result.FileExcluder = file.MatchAppleProtected(result.FileExcluder)
+		result.DirExcluder = file.MatchAppleProtected(result.DirExcluder)
+	}
+
 	return result, nil
 }

--- a/go.mod
+++ b/go.mod
@@ -3,7 +3,7 @@ module github.com/andrejacobs/ajfs
 go 1.25
 
 require (
-	github.com/andrejacobs/go-aj v0.2.0
+	github.com/andrejacobs/go-aj v0.2.1
 	github.com/andrejacobs/go-collection v0.1.0
 	github.com/schollz/progressbar/v3 v3.19.0
 	github.com/spf13/cobra v1.10.2

--- a/internal/app/scan/scan.go
+++ b/internal/app/scan/scan.go
@@ -113,6 +113,10 @@ func Run(cfg Config) error {
 			if err := dbf.Close(); err != nil {
 				fmt.Fprintln(cfg.Stderr, err)
 			}
+		} else {
+			// Close file and remove it since it is damaged
+			cfg.Errorln("\nApp was interrupted and the ajfs database file is incomplete. File will be deleted.")
+			_ = dbf.Interrupted()
 		}
 	}()
 
@@ -142,9 +146,7 @@ func Run(cfg Config) error {
 
 	cfg.ProgressPrintln("Scanning ...")
 	if err = s.Scan(ctx, dbf); err != nil {
-		if !errors.Is(err, context.Canceled) {
-			return err
-		}
+		return err
 	}
 
 	safeToShutdown = true

--- a/internal/app/scan/scan.go
+++ b/internal/app/scan/scan.go
@@ -58,7 +58,8 @@ type Config struct {
 	DryRun   bool // Only display files and directories that would have been stored in the database.
 	InitOnly bool // The initial database will be created without long running processes (hashing).
 
-	simulateHashingError bool // Cause an error while calculating file signature hashes.
+	simulateScanningError bool // Cause an error while scanning.
+	simulateHashingError  bool // Cause an error while calculating file signature hashes.
 }
 
 // The hashing function to be used for calculating file signature hashes.
@@ -104,9 +105,14 @@ func Run(cfg Config) error {
 		return err
 	}
 
+	safeToShutdown := false
+
 	defer func() {
-		if err := dbf.Close(); err != nil {
-			fmt.Fprintln(cfg.Stderr, err)
+		if safeToShutdown {
+			// Only close and verify if we did not encounter an error during the scanning process
+			if err := dbf.Close(); err != nil {
+				fmt.Fprintln(cfg.Stderr, err)
+			}
 		}
 	}()
 
@@ -117,8 +123,6 @@ func Run(cfg Config) error {
 	signalCh := make(chan os.Signal, 1)
 	interruptedCh := make(chan bool, 1)
 	signal.Notify(signalCh, os.Interrupt, syscall.SIGTERM)
-
-	safeToShutdown := false
 
 	go func() {
 		rcv := <-signalCh
@@ -141,8 +145,20 @@ func Run(cfg Config) error {
 		if !errors.Is(err, context.Canceled) {
 			return err
 		}
-	} else {
-		safeToShutdown = true
+	}
+
+	safeToShutdown = true
+
+	if cfg.simulateScanningError {
+		if err := dbf.StartHashTable(cfg.Algo); err != nil {
+			return err
+		}
+
+		if err := dbf.FinishHashTable(); err != nil {
+			return err
+		}
+
+		return fmt.Errorf("simulating an error while scanning")
 	}
 
 	if cfg.CalculateHashes && (ctx.Err() == nil) {

--- a/internal/app/scan/scan_priv_test.go
+++ b/internal/app/scan/scan_priv_test.go
@@ -41,6 +41,27 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
+func TestScanWithErrors(t *testing.T) {
+	tempFile := filepath.Join(t.TempDir(), "unit-testing")
+	_ = os.Remove(tempFile)
+	defer os.Remove(tempFile)
+
+	cfg := initialConfig()
+	cfg.DbPath = tempFile
+	cfg.CalculateHashes = true
+	cfg.Algo = ajhash.AlgoSHA1
+
+	// Cause an error while scanning
+	cfg.simulateScanningError = true
+
+	var err error
+	require.NotPanics(t, func() {
+		err = Run(cfg)
+	})
+
+	require.ErrorContains(t, err, "simulating an error while scanning")
+}
+
 func TestScanWithHashingErrors(t *testing.T) {
 	tempFile := filepath.Join(t.TempDir(), "unit-testing")
 	_ = os.Remove(tempFile)

--- a/internal/scanner/scanner.go
+++ b/internal/scanner/scanner.go
@@ -26,6 +26,7 @@ import (
 	"fmt"
 	"io/fs"
 	"path/filepath"
+	"runtime"
 
 	"github.com/andrejacobs/ajfs/internal/db"
 	"github.com/andrejacobs/ajfs/internal/path"
@@ -54,6 +55,10 @@ func NewScanner() Scanner {
 
 // Return the default file excluder.
 func DefaultFileExcluder() file.MatchPathFn {
+	if runtime.GOOS == "darwin" {
+		return file.MatchAppleProtected(file.MatchAppleDSStore(file.MatchNever))
+	}
+
 	return file.MatchAppleDSStore(file.MatchNever)
 }
 

